### PR TITLE
Preallocated Contexts - Take 2

### PR DESCRIPTION
**Pull request by DavidBurkett**
Replaces https://github.com/mimblewimble/secp256k1-zkp/pull/61

Slightly more focused, with lower risk

### CODE REVIEW

**Comment by phyro** [comment]
@GeneFerneau could you throw an eye again on this one?
Would it make sense to get this one merged as well https://github.com/mimblewimble/secp256k1-zkp/pull/62/files and rerun the tests?

File: src/ecmult_gen_impl.h [hunk 1]
```diff
@@ -7,30 +7,40 @@
...
+    size_t const prealloc_size = SECP256K1_ECMULT_GEN_CONTEXT_PREALLOCATED_SIZE;
+    void* const base = *prealloc;
 #endif
 
     if (ctx->prec != NULL) {
         return;
     }
 #ifndef USE_ECMULT_STATIC_PRECOMPUTATION
-    ctx->prec = (secp256k1_ge_storage (*)[64][16])checked_malloc(cb, sizeof(*ctx->prec));
+    ctx->prec = (secp256k1_ge_storage (*)[64][16])manual_alloc(prealloc, prealloc_size, base, prealloc_size);
```
Comment (quentinlesceller) [inline src/ecmult_gen_impl.h:43]: Original repo doesn't have this `[64][16]`. Any idea why?

https://github.com/bitcoin-core/secp256k1/blob/1e5d50fa93d71d751b95eec6a80f6732879a0071/src/ecmult_gen_impl.h#L43

File: src/ecmult_gen_impl.h [hunk 2]
```diff
@@ -95,27 +105,18 @@ static int secp256k1_ecmult_gen_context_is_built(const secp256k1_ecmult_gen_cont
...
-    if (src->prec == NULL) {
-        dst->prec = NULL;
-    } else {
+static void secp256k1_ecmult_gen_context_finalize_memcpy(secp256k1_ecmult_gen_context *dst, const secp256k1_ecmult_gen_context *src) {
 #ifndef USE_ECMULT_STATIC_PRECOMPUTATION
-        dst->prec = (secp256k1_ge_storage (*)[64][16])checked_malloc(cb, sizeof(*dst->prec));
-        memcpy(dst->prec, src->prec, sizeof(*dst->prec));
+    if (src->prec != NULL) {
+        /* We cast to void* first to suppress a -Wcast-align warning. */
+        dst->prec = (secp256k1_ge_storage (*)[64][16])(void*)((unsigned char*)dst + ((unsigned char*)src->prec - (unsigned char*)src));
```
Comment (quentinlesceller) [inline src/ecmult_gen_impl.h:112]: Same comment here.

File: src/ecmult_impl.h [hunk 3]
```diff
@@ -137,26 +138,110 @@ static void secp256k1_ecmult_odd_multiples_table_globalz_windowa(secp256k1_ge *p
...
-    secp256k1_ge *prea = (secp256k1_ge*)checked_malloc(cb, sizeof(secp256k1_ge) * n);
-    secp256k1_fe *zr = (secp256k1_fe*)checked_malloc(cb, sizeof(secp256k1_fe) * n);
+static void secp256k1_ecmult_odd_multiples_table_storage_var(const int n, secp256k1_ge_storage *pre, const secp256k1_gej *a) {
+    secp256k1_gej d;
+    secp256k1_ge a_ge, d_ge, p_ge;
+    secp256k1_ge last_ge;
+    secp256k1_gej pj;
+    secp256k1_fe zi;
+    secp256k1_fe zr;
+    secp256k1_fe dx_over_dz_squared;
```
Comment (quentinlesceller) [inline src/ecmult_impl.h:148]: This function is also different from the original repo. 
https://github.com/bitcoin-core/secp256k1/blob/1e5d50fa93d71d751b95eec6a80f6732879a0071/src/ecmult_impl.h#L145

File: src/ecmult_gen_impl.h (hunk 1 above)
Comment (DavidBurkett) [reply src/ecmult_gen_impl.h:43]: The [64][16] was replaced by a macro as part of this commit: https://github.com/bitcoin-core/secp256k1/commit/dcb2e3b3fff0b287d576842aabe5c79f2fe4df30

We'll eventually get to that as well.

Comment (quentinlesceller) [reply src/ecmult_gen_impl.h:43]: Okay sounds good.

File: src/ecmult_impl.h [hunk 4]
```diff
@@ -137,26 +138,110 @@ static void secp256k1_ecmult_odd_multiples_table_globalz_windowa(secp256k1_ge *p
...
+     * `secp256k1_gej_add_ge_var` to perform the additions. For each addition, we store
+     * the resulting y-coordinate and the z-ratio, since we only have enough memory to
+     * store two field elements. These are sufficient to efficiently undo the isomorphism
+     * and recompute all the `x`s.
+     */
+    d_ge.x = d.x;
+    d_ge.y = d.y;
+    d_ge.infinity = 0;
+
+    secp256k1_ge_set_gej_zinv(&a_ge, a, &d.z);
```
Comment (quentinlesceller) [inline src/ecmult_impl.h:166]: Slight variation here. But maybe `secp256k1_ge_set_gej_zinv` interface changed.

**Review by quentinlesceller — APPROVED** [review]
Looks good to me. Cross checked every line with:
- https://github.com/bitcoin-core/secp256k1
- https://github.com/ElementsProject/secp256k1-zkp
Specifically regarding the differences  in `secp256k1_ecmult_odd_multiples_table_storage_var` compared with commit at https://github.com/ElementsProject/secp256k1-zkp/commit/efa783f8f0addb8f1e2ee0f1b9999673c6414acd

File: src/secp256k1.c [hunk 5]
```diff
@@ -84,45 +84,113 @@ static const secp256k1_context secp256k1_context_no_precomp_ = {
...
+
+    VERIFY_CHECK(ctx != NULL);
+    prealloc_size = secp256k1_context_preallocated_clone_size(ctx);
+    ret = (secp256k1_context*)checked_malloc(&ctx->error_callback, prealloc_size);
+    ret = secp256k1_context_preallocated_clone(ctx, ret);
     return ret;
 }
 
-void secp256k1_context_destroy(secp256k1_context* ctx) {
+void secp256k1_context_preallocated_destroy(secp256k1_context* ctx) {
```
[the remark this answers is not in the document]
Comment (DavidBurkett) [reply src/secp256k1.c:184]: That was added as part of a separate code change (https://github.com/bitcoin-core/secp256k1/commit/40fde611bd145bdafb78a52af6d197489101bafb which was later replaced by https://github.com/bitcoin-core/secp256k1/commit/6095a863faf68ee41ddf510b0f6b05a45ec6efc3).

We can add those changes in a separate PR.

File: src/secp256k1.c [hunk 6]
```diff
@@ -84,45 +84,113 @@ static const secp256k1_context secp256k1_context_no_precomp_ = {
...
+    }
+    return ret;
+}
+
+secp256k1_context* secp256k1_context_preallocated_create(void* prealloc, unsigned int flags) {
+    void* const base = prealloc;
+    size_t prealloc_size;
+    secp256k1_context* ret;
+
+    VERIFY_CHECK(prealloc != NULL);
```
[the remark this answers is not in the document]
Comment (DavidBurkett) [reply src/secp256k1.c:124]: This change was made fairly recently, in this commit: https://github.com/bitcoin-core/secp256k1/commit/ebfa2058e9cc2999dada47d2f1e1e5c0f4bcf619

We can make that change in a separate PR.

**Pull request merged**